[SPARK-14388][SQL] Implement CREATE TABLE

andrewor14 opened the pull request:
## What changes were proposed in this pull request?

This patch implements the `CREATE TABLE` command using the `SessionCatalog`. Previously we handled only `CTAS` and `CREATE TABLE ... USING`. This requires us to refactor `CatalogTable` to accept various fields (e.g. bucket and skew columns) and pass them to Hive.

WIP: Note that I haven't verified whether this actually works yet! But I believe it does.
## How was this patch tested?

Tests will come in a future commit.

Commit:
Refactor CatalogTable column semantics

Before: CatalogTable has schema, partitionColumns and sortColumns.
There are no constraints between the 3. However, Hive will
complain if schema and partitionColumns overlap.

After: CatalogTable has schema, partitionColumnNames,
sortColumnNames, bucketColumnNames and skewColumnNames. All the
columns must be a subset of schema. This means splitting up
schema into (schema, partitionCols) before passing it to Hive.

This allows us to store the columns more uniformly. Otherwise
partition columns would be the odd one out. This commit also
fixes "alter table bucketing", which was incorrectly using
partition columns as bucket columns.

Commit:
Parse various parts of the CREATE TABLE command

We need to reconcile the differences between what's added here in
SparkSqlParser and HiveSqlParser. That will come in the next
commit.

This currently still fails tests, obviously because create table
is not implemented yet!

Commit:
Implement CREATE TABLE in Hive parser

This involves reverting part of the changes in an earlier commit,
where we tried to implement the parsing logic in the general SQL
parser and introduced a bunch of case classes that we won't end
up using.

As of this commit the actual CREATE TABLE logic is not there yet.
It will come in a future commit.

Commit:
Implement it

SparkQA commented:
**[Test build #55425 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/55425/consoleFull)** for PR 12271 at commit [`5e0fe03`](https://github.com/apache/spark/commit/5e0fe03bfa655c6de854bc8adaa73186a17a0b0c).

andrewor14 commented on sql/core/src/main/scala/org/apache/spark/sql/execution/command/ddl.scala:
```diff
@@ -176,29 +176,6 @@ case class DescribeDatabase(
- *    ALTER VIEW view1 RENAME TO view2;
- * }}}
- */
-case class AlterTableRename(
```
moved to `tables.scala`. I just moved this one command for now to avoid inflating the diff too much.

andrewor14 commented on sql/core/src/test/scala/org/apache/spark/sql/execution/command/DDLSuite.scala:
```diff
@@ -320,8 +320,6 @@ class DDLSuite extends QueryTest with SharedSQLContext with BeforeAndAfterEach {
     assertUnsupported("ALTER TABLE dbx.tab1 NOT STORED AS DIRECTORIES")
   }
 
-  // TODO: ADD a testcase for Drop Database in Restric when we can create tables in SQLContext
```
this TODO comment is duplicated in line 197

Commit:
Revert unnecessary changes (small)

andrewor14 commented on sql/catalyst/src/main/antlr4/org/apache/spark/sql/catalyst/parser/SqlBase.g4:
```diff
@@ -271,8 +271,7 @@ createFileFormat
 
 fileFormat
-    : INPUTFORMAT inFmt=STRING OUTPUTFORMAT outFmt=STRING (SERDE serdeCls=STRING)?
-      (INPUTDRIVER inDriver=STRING OUTPUTDRIVER outDriver=STRING)?                         #tableFileFormat
```
@hvanhovell I deleted this because Hive doesn't support it. Why was this added in the first place? Is there any supporting documentation for this somewhere?

hvanhovell replied:
@andrewor14 I wanted to make sure we supported the same grammar as Hive and I used their grammars as a basis. So this is defined in the following two locations:

- https://github.com/apache/hive/blob/master/ql/src/java/org/apache/hadoop/hive/ql/parse/HiveParser.g#L1335-L1336
- https://github.com/apache/hive/blob/master/ql/src/java/org/apache/hadoop/hive/ql/parse/HiveParser.g#L1335-L1336

The main idea was that I could throw better errors. But if it is not supported by Hive itself then please remove it!

yhuai replied:
Let's remove it. I have never seen this before.

SparkQA commented:
**[Test build #55426 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/55426/consoleFull)** for PR 12271 at commit [`f7501d9`](https://github.com/apache/spark/commit/f7501d9ebc5c4f08374788a937de6a56689258b8).

SparkQA commented:
**[Test build #55425 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/55425/consoleFull)** for PR 12271 at commit [`5e0fe03`](https://github.com/apache/spark/commit/5e0fe03bfa655c6de854bc8adaa73186a17a0b0c).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/55425/
Test FAILed.

SparkQA commented:
**[Test build #55426 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/55426/consoleFull)** for PR 12271 at commit [`f7501d9`](https://github.com/apache/spark/commit/f7501d9ebc5c4f08374788a937de6a56689258b8).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/55426/
Test FAILed.

xwu0226 commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/catalog/interface.scala:
```diff
@@ -218,14 +219,35 @@ case class CatalogTable(
     schema: Seq[CatalogColumn],
-    partitionColumns: Seq[CatalogColumn] = Seq.empty,
-    sortColumns: Seq[CatalogColumn] = Seq.empty,
+    partitionColumnNames: Seq[String] = Seq.empty,
```
The hive syntax has PARTITIONED BY (col_name  data_type  [COMMENT col_comment], ...). The column definition here is the same as the ones in schema. CatalogColumn can represent these 3 aspects of each column.. why we need to change to use String here?

gatorsmile commented:
@andrewor14 Just want to let you know, @xwu0226 is doing the command `SHOW CREATE TABLE`. He is writing many test cases in the PR: https://github.com/apache/spark/pull/12132/files. It might help you in this PR. Thanks!

yhuai commented on sql/core/src/main/scala/org/apache/spark/sql/execution/command/tables.scala:
```diff
@@ -0,0 +1,80 @@
+ *   [COMMENT table_comment]
+ *   [PARTITIONED BY (col3 data_type [COMMENT col_comment], ...)]
+ *   [CLUSTERED BY (col1, ...) [SORTED BY (col1 [ASC|DESC], ...)] INTO num_buckets BUCKETS]
+ *   [SKEWED BY (col1, col2, ...) ON ((col_value, col_value, ...), ...)
```
@andrewor14 I just did a quick check with our `InsertIntoHiveTable` command. Seems this command does not really understand how to handle a table having specifications on `CLUSTERED BY`, `SORTED BY` or `SKEWED BY`. How about we just throw exceptions when a define provide these specs?

andrewor14 replied:
ok

yhuai commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/catalog/interface.scala:
```diff
@@ -218,14 +219,35 @@ case class CatalogTable(
+    require(cols.toSet.subsetOf(colNames), s"$colType columns (${cols.mkString(", ")}) " +
+      s"must be a subset of schema (${colNames.mkString(", ")}) in table '$identifier'")
+  }
+  requireSubsetOfSchema(partitionColumnNames, "partition")
```
I find the following example from Hive's doc
```
CREATE TABLE page_view(viewTime INT, userid BIGINT,
     page_url STRING, referrer_url STRING,
     ip STRING COMMENT 'IP Address of the User')
 COMMENT 'This is the page view table'
 PARTITIONED BY(dt STRING, country STRING)
 ROW FORMAT DELIMITED
   FIELDS TERMINATED BY '\001'
STORED AS SEQUENCEFILE;
```

Will this CREATE TABLE statement trigger an error at here?

andrewor14 replied:
no because we put the partition columns in the schema when we parse it. There's a comment in `HiveSqlParser` that describes this.

yhuai replied:
Thanks!

yhuai commented on sql/hive/src/main/scala/org/apache/spark/sql/hive/client/HiveClientImpl.scala:
```diff
@@ -665,9 +672,17 @@ private[hive] class HiveClientImpl(
     })
-    hiveTable.setFields(table.schema.map(toHiveColumn).asJava)
-    hiveTable.setPartCols(table.partitionColumns.map(toHiveColumn).asJava)
+    // TODO: pass the comment to Hive somehow
```
comment is stored in the table properties with a field called `comment`.

andrewor14 replied:
I see. That's what I suspected

hvanhovell commented on sql/hive/src/main/scala/org/apache/spark/sql/hive/execution/HiveSqlParser.scala:
```diff
@@ -134,82 +136,117 @@ class HiveSqlAstBuilder extends SparkSqlAstBuilder {
+        }
+      case (null, storageHandler) if storageHandler != null =>
+        throw new ParseException("Operation not allowed: ... STORED BY storage_handler ...", ctx)
+      case (null, null) =>
```
I don't think this is possible. If they are both empty, the `CreateFileFormatContext` should be null to begin with.

andrewor14 replied:
sure

hvanhovell commented on sql/hive/src/main/scala/org/apache/spark/sql/hive/execution/HiveSqlParser.scala:
```diff
@@ -134,82 +136,117 @@ class HiveSqlAstBuilder extends SparkSqlAstBuilder {
-      typedVisit[CatalogStorageFormat](ctx.fileFormat)
-    } else {
-      visitStorageHandler(ctx.storageHandler)
+    (ctx.fileFormat, ctx.storageHandler) match {
```
The parser either allows a fileFormat or a storageHandler here (not both). So we could simplify this.

hvanhovell commented on sql/hive/src/main/scala/org/apache/spark/sql/hive/execution/HiveSqlParser.scala:
```diff
@@ -397,11 +429,28 @@ class HiveSqlAstBuilder extends SparkSqlAstBuilder {
-      ctx: StorageHandlerContext): CatalogStorageFormat = withOrigin(ctx) {
-    throw new ParseException("Storage Handlers are currently unsupported.", ctx)
+  private def visitRowFormat(ctx: RowFormatContext): CatalogStorageFormat = withOrigin(ctx) {
+    ctx match {
```
Minor: You could also call `ctx.accept(this)` here and cast it, but I get that this is more typed and less opaque.

andrewor14 replied:
yeah, that's the whole reason why I changed it. There's too much magic going on in `typedAccept` or whatever it's called, especially in code that is generated and hidden from the IDE.

hvanhovell commented on sql/hive/src/main/scala/org/apache/spark/sql/hive/execution/HiveSqlParser.scala:
```diff
@@ -134,82 +136,117 @@ class HiveSqlAstBuilder extends SparkSqlAstBuilder {
+   *   [AS select_statement];
+   * }}}
    */
   override def visitCreateTable(ctx: CreateTableContext): LogicalPlan = {
```
Could you add withOrigin here? I apparently forgot to...

andrewor14 replied:
ok

hvanhovell commented on sql/hive/src/main/scala/org/apache/spark/sql/hive/execution/HiveSqlParser.scala:
```diff
@@ -134,82 +136,117 @@ class HiveSqlAstBuilder extends SparkSqlAstBuilder {
+    val (name, temp, ifNotExists, external) = visitCreateTableHeader(ctx.createTableHeader)
+    // TODO: implement temporary tables
+    if (temp) {
+      throw new AnalysisException(
```
I would throw a ParseException here. The advantage is that the error reporting of these exceptions is a bit better.

andrewor14 replied:
I'll address all of these in SPARK-14414. For now I just made this (and a few others) an `AnalysisException` so we can reuse some code in testing. I'll fix this before 2.0.

andrewor14 replied:
(I actually fixed this in this patch, but we still need to fix it in all the other places in SPARK-14414)

xwu0226 commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/catalog/interface.scala:
```diff
@@ -218,14 +219,35 @@ case class CatalogTable(
-    partitionColumns: Seq[CatalogColumn] = Seq.empty,
-    sortColumns: Seq[CatalogColumn] = Seq.empty,
+    partitionColumnNames: Seq[String] = Seq.empty,
+    sortColumnNames: Seq[String] = Seq.empty,
```
How do we maintain the ordering of the sort columns here? For the clause SORTED BY (col_name [ASC|DESC], ...), within CLUSTERED BY clause.

yhuai replied:
I think we do not support SORTED BY and CLUSTERED BY for now.

andrewor14 replied:
also that has nothing to do with this patch

yhuai commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/catalog/interface.scala:
```diff
@@ -218,14 +219,35 @@ case class CatalogTable(
     viewOriginalText: Option[String] = None,
-    viewText: Option[String] = None) {
+    viewText: Option[String] = None,
+    comment: Option[String] = None) {
```
Maybe renaming it to `description` (be consistent with `CatalogDatabase`)?

xwu0226 replied:
Does this "comment" mean the comment for the table, as defined in the syntax `create table t1 (c1 int) COMMENT 'abc'`?  If it is, shall it be put into `properties`, it seems Hive puts this table comment into TBLPROPERTIES. so it is consistent when creating hive metastore table from CatalogTable object?

andrewor14 replied:
I actually prefer to keep it as comment because it's `COMMENT` in the query and `"comment"` in Hive's table properties

yhuai replied:
ok

Commit:
Merge branch 'master' of github.com:apache/spark into create-table-ddl

Commit:
Address comment

Commit:
Add all the tests

SparkQA commented:
**[Test build #55552 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/55552/consoleFull)** for PR 12271 at commit [`2e95ecf`](https://github.com/apache/spark/commit/2e95ecf790dc5d5b12b6ec72c0bd2b4bca99b17d).

SparkQA commented:
**[Test build #55552 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/55552/consoleFull)** for PR 12271 at commit [`2e95ecf`](https://github.com/apache/spark/commit/2e95ecf790dc5d5b12b6ec72c0bd2b4bca99b17d).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/55552/
Test FAILed.

Commit:
Merge branch 'master' of github.com:apache/spark into create-table-ddl

yhuai commented on sql/hive/src/main/scala/org/apache/spark/sql/hive/execution/HiveSqlParser.scala:
```diff
@@ -121,84 +123,114 @@ class HiveSqlAstBuilder extends SparkSqlAstBuilder {
+   *   [COMMENT table_comment]
+   *   [PARTITIONED BY (col3 data_type [COMMENT col_comment], ...)]
+   *   [CLUSTERED BY (col1, ...) [SORTED BY (col1 [ASC|DESC], ...)] INTO num_buckets BUCKETS]
+   *   [SKEWED BY (col1, col2, ...) ON ((col_value, col_value, ...), ...) [STORED AS DIRECTORIES]]
```
Remove this line while resolving the conflicts?

yhuai commented on sql/hive/src/main/scala/org/apache/spark/sql/hive/execution/HiveSqlParser.scala:
```diff
@@ -121,84 +123,114 @@ class HiveSqlAstBuilder extends SparkSqlAstBuilder {
-      HiveNativeCommand(command(ctx))
+  override def visitCreateTable(ctx: CreateTableContext): LogicalPlan = withOrigin(ctx) {
+    val (name, temp, ifNotExists, external) = visitCreateTableHeader(ctx.createTableHeader)
+    // TODO: implement temporary tables
```
+1

yhuai commented on sql/hive/src/main/scala/org/apache/spark/sql/hive/execution/HiveSqlParser.scala:
```diff
@@ -121,84 +123,114 @@ class HiveSqlAstBuilder extends SparkSqlAstBuilder {
+      serde = rowStorage.serde.orElse(fileStorage.serde).orElse(defaultStorage.serde),
+      serdeProperties = rowStorage.serdeProperties ++ fileStorage.serdeProperties)
+
+    // TODO support the sql text - have a proper location for this!
```
Maybe explain this?

andrewor14 replied:
@hvanhovell maybe explain this? :)

hvanhovell replied:
In the old situation we were storing the `comment` into the `viewText` column which didn't seem right. I think we can remove this one now that we actually have a comment field.

yhuai commented:
LGTM

Commit:
Not OK

andrewor14 commented:
Thanks for the reviews. Latest commit should pass tests.

SparkQA commented:
**[Test build #55628 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/55628/consoleFull)** for PR 12271 at commit [`250f402`](https://github.com/apache/spark/commit/250f402372e9826865749f3b81cd96a7cdaff657).

SparkQA commented:
**[Test build #55628 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/55628/consoleFull)** for PR 12271 at commit [`250f402`](https://github.com/apache/spark/commit/250f402372e9826865749f3b81cd96a7cdaff657).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/55628/
Test FAILed.

yhuai commented:
test this please

SparkQA commented:
**[Test build #55634 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/55634/consoleFull)** for PR 12271 at commit [`250f402`](https://github.com/apache/spark/commit/250f402372e9826865749f3b81cd96a7cdaff657).

hvanhovell commented:
LGTM - pending jenkins

SparkQA commented:
**[Test build #55634 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/55634/consoleFull)** for PR 12271 at commit [`250f402`](https://github.com/apache/spark/commit/250f402372e9826865749f3b81cd96a7cdaff657).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/55634/
Test FAILed.

andrewor14 commented:
retest this please

SparkQA commented:
**[Test build #55644 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/55644/consoleFull)** for PR 12271 at commit [`250f402`](https://github.com/apache/spark/commit/250f402372e9826865749f3b81cd96a7cdaff657).

SparkQA commented:
**[Test build #55644 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/55644/consoleFull)** for PR 12271 at commit [`250f402`](https://github.com/apache/spark/commit/250f402372e9826865749f3b81cd96a7cdaff657).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/55644/
Test FAILed.

Commit:
Fix part of InsertIntoHiveTableSuite

We weren't using the right default serde in Hive. Note that this
still fails a test with "Reference 'ds' is ambiguous ...", but
this error is common across many tests so it will be addressed
in a future commit.

SparkQA commented:
**[Test build #55654 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/55654/consoleFull)** for PR 12271 at commit [`efecac9`](https://github.com/apache/spark/commit/efecac9b01b3ff8be296234392f4a6c922fa2d25).

Commit:
Fix ambiguous reference bug

In HiveMetastoreCatalog we already combined the schema and the
partition keys to compensate for the fact that Hive separates it.
Now this logic is pushed to the edges where Spark talks to Hive.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/55654/
Test FAILed.

SparkQA commented:
**[Test build #55658 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/55658/consoleFull)** for PR 12271 at commit [`50a2054`](https://github.com/apache/spark/commit/50a2054ec7a7276d45c1ab5adabd4550e00c7811).

Commit:
Fix ParquetMetastoreSuite

Previously we always converted the data type string to lower case.
However, for struct fields this also converts the struct field
names to lower case. This is not what tests (or perhaps user code)
expects.

SparkQA commented:
**[Test build #55666 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/55666/consoleFull)** for PR 12271 at commit [`8dc554a`](https://github.com/apache/spark/commit/8dc554a38c9989fc43b119645bfe5c8ceb7b6cdb).

Commit:
Fix SQLQuerySuite

SparkQA commented:
**[Test build #55658 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/55658/consoleFull)** for PR 12271 at commit [`50a2054`](https://github.com/apache/spark/commit/50a2054ec7a7276d45c1ab5adabd4550e00c7811).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/55658/
Test FAILed.

SparkQA commented:
**[Test build #55666 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/55666/consoleFull)** for PR 12271 at commit [`8dc554a`](https://github.com/apache/spark/commit/8dc554a38c9989fc43b119645bfe5c8ceb7b6cdb).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/55666/
Test FAILed.

Commit:
Fix HiveCompatibilitySuite (ignored some tests)

Commit:
Fix HiveDDLCommandSuite

andrewor14 commented:
I've ignored some tests in `HiveCompatibilitySuite` for now. I'll have a look at them shortly. I just wanted to see if the rest of the tests will pass.

SparkQA commented:
**[Test build #55674 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/55674/consoleFull)** for PR 12271 at commit [`8e273fd`](https://github.com/apache/spark/commit/8e273fdc4f95d08cb6d09f4641472861587a3a01).

SparkQA commented:
**[Test build #55674 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/55674/consoleFull)** for PR 12271 at commit [`8e273fd`](https://github.com/apache/spark/commit/8e273fdc4f95d08cb6d09f4641472861587a3a01).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/55674/
Test FAILed.

yhuai commented on sql/hive/compatibility/src/test/scala/org/apache/spark/sql/hive/execution/HiveCompatibilitySuite.scala:
```diff
@@ -509,7 +555,7 @@ class HiveCompatibilitySuite extends HiveQueryFileTest with BeforeAndAfter {
     "date_join1",
     "date_serde",
-    "decimal_1",
+    //"decimal_1", // TODO: cannot parse column decimal(5)
```
Yea. We should support decimal(5). At here, the scale is 0.

yhuai replied:
https://issues.apache.org/jira/browse/SPARK-14591

yhuai commented on sql/hive/compatibility/src/test/scala/org/apache/spark/sql/hive/execution/HiveCompatibilitySuite.scala:
```diff
@@ -490,15 +538,13 @@ class HiveCompatibilitySuite extends HiveQueryFileTest with BeforeAndAfter {
     "cp_mj_rc",
     "create_insert_outputformat",
-    "create_like_tbl_props",
+    //"create_like_tbl_props", // TODO: results don't match
```
https://issues.apache.org/jira/browse/SPARK-14592

yhuai commented on sql/hive/compatibility/src/test/scala/org/apache/spark/sql/hive/execution/HiveCompatibilitySuite.scala:
```diff
@@ -790,14 +821,13 @@ class HiveCompatibilitySuite extends HiveQueryFileTest with BeforeAndAfter {
     "parenthesis_star_by",
     "part_inherit_tbl_props",
-    "part_inherit_tbl_props_empty",
+    //"part_inherit_tbl_props_empty", // TODO: results don't match
```
This one and others that are commented out below are caused by differences of describe table. There are mainly three differences. First, for managed tables, we will EXTERNAL set to false in table properties. For tables that are not bucketed, we will have numBuckets set to 0 instead of -1. For tables that does not specify the file format, the output format will be org.apache.hadoop.hive.ql.io.IgnoreKeyTextOutputFormat instead of org.apache.hadoop.hive.ql.io.HiveIgnoreKeyTextOutputFormat (For Hive, it somehow replaced even if the explain output still shows IgnoreKeyTextOutputFormat). We do not need to block this PR for any of them.

yhuai commented on sql/hive/src/main/scala/org/apache/spark/sql/hive/execution/HiveSqlParser.scala:
```diff
@@ -121,84 +123,115 @@ class HiveSqlAstBuilder extends SparkSqlAstBuilder {
+        outputFormat = defaultHiveSerde.flatMap(_.outputFormat)
+          .orElse(Some("org.apache.hadoop.hive.ql.io.IgnoreKeyTextOutputFormat")),
+        serde = defaultHiveSerde.flatMap(_.serde)
+          .orElse(Some("org.apache.hadoop.hive.serde2.lazy.LazySimpleSerDe")),
```
This one breaks the CTAS test.

yhuai replied:
I guess we should fill this in as late as possible? Probably in `toHiveTable`? In `HiveClientImpl.toHiveTable`, we first create a `org.apache.hadoop.hive.ql.metadata.Table`, which uses `MetadataTypedColumnsetSerDe` as the default one (so, originally `InsertIntoHiveTableSuite` was broken without this change).

Commit:
Fix SQLQuerySuite CTAS

yhuai commented on sql/hive/src/main/scala/org/apache/spark/sql/hive/execution/HiveSqlParser.scala:
```diff
@@ -435,13 +480,15 @@ class HiveSqlAstBuilder extends SparkSqlAstBuilder {
+        // just convert the whole type string to lower case, otherwise the struct field names
+        // will no longer be case sensitive. Instead, we rely on our parser to get the proper
+        // case before passing it to Hive.
+        HiveMetastoreTypes.toDataType(col.dataType.getText).simpleString,
```
I think we need to use `parseDataType`.

SparkQA commented:
**[Test build #55692 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/55692/consoleFull)** for PR 12271 at commit [`59edce3`](https://github.com/apache/spark/commit/59edce332f87b07bdfb07e2e385431b2b123e1b0).

Commit:
Fix all but 1 ignored test in HiveCompatibilitySuite

There were a few differences in DESCRIBE TABLE:
- output format should be HiveIgnoreKeyTextOutputFormat
- num buckets should be -1
- last access time should be -1
- EXTERNAL should not be set to false for managed table

After making these changes out result now matches Hive's.

Commit:
Fix last ignored test in HiveCompatibilitySuite

CatalystSqlParser knows how to parse decimal(5)!

Commit:
Merge branch 'master' of github.com:apache/spark into create-table-ddl

SparkQA commented:
**[Test build #55693 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/55693/consoleFull)** for PR 12271 at commit [`a60e66a`](https://github.com/apache/spark/commit/a60e66a71dec96973043ec62e2b6d4213c5add2c).

SparkQA commented:
**[Test build #55696 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/55696/consoleFull)** for PR 12271 at commit [`02738fe`](https://github.com/apache/spark/commit/02738fed54ba0eebc2c8d887430f0bef34213c68).

hvanhovell commented on sql/hive/src/main/scala/org/apache/spark/sql/hive/execution/HiveSqlParser.scala:
```diff
@@ -488,7 +488,7 @@ class HiveSqlAstBuilder extends SparkSqlAstBuilder {
         // will no longer be case sensitive. Instead, we rely on our parser to get the proper
         // case before passing it to Hive.
-        HiveMetastoreTypes.toDataType(col.dataType.getText).simpleString,
+        CatalystSqlParser.parseDataType(col.dataType.getText).simpleString,
```
MINOR/NIT: The DataType parsing is done in the `AstBuilder`, so we really don't need to parse this string again. You could use (magical/evil) `typedVisit[DataType](col.dataType)` here.

hvanhovell commented on sql/hive/src/main/scala/org/apache/spark/sql/hive/execution/HiveSqlParser.scala:
```diff
@@ -435,13 +480,15 @@ class HiveSqlAstBuilder extends SparkSqlAstBuilder {
+        // just convert the whole type string to lower case, otherwise the struct field names
+        // will no longer be case sensitive. Instead, we rely on our parser to get the proper
+        // case before passing it to Hive.
+        CatalystSqlParser.parseDataType(col.dataType.getText).simpleString,
```
MINOR/NIT: The DataType parsing is done in the `AstBuilder`, so we really don't need to parse this string again. You could use (magical/evil) `typedVisit[DataType](col.dataType)` here.

SparkQA commented:
**[Test build #55692 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/55692/consoleFull)** for PR 12271 at commit [`59edce3`](https://github.com/apache/spark/commit/59edce332f87b07bdfb07e2e385431b2b123e1b0).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/55692/
Test FAILed.

SparkQA commented:
**[Test build #55693 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/55693/consoleFull)** for PR 12271 at commit [`a60e66a`](https://github.com/apache/spark/commit/a60e66a71dec96973043ec62e2b6d4213c5add2c).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/55693/
Test FAILed.

SparkQA commented:
**[Test build #55696 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/55696/consoleFull)** for PR 12271 at commit [`02738fe`](https://github.com/apache/spark/commit/02738fed54ba0eebc2c8d887430f0bef34213c68).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/55696/
Test FAILed.

yhuai commented on sql/hive/src/main/scala/org/apache/spark/sql/hive/client/HiveClientImpl.scala:
```diff
@@ -675,34 +680,38 @@ private[hive] class HiveClientImpl(
     table.storage.outputFormat.map(toOutputFormat).foreach(hiveTable.setOutputFormatClass)
-    table.storage.serde.foreach(hiveTable.setSerializationLib)
+    hiveTable.setSerializationLib(
+      table.storage.serde.getOrElse("org.apache.hadoop.hive.serde2.lazy.LazySimpleSerDe"))
```
hmm... Without this PR, the storage related fields of the table defined in `org.apache.spark.sql.hive.MetastoreDataSourcesSuite.persistent JSON table` are:
```
[SerDe Library:      	org.apache.hadoop.hive.serde2.MetadataTypedColumnsetSerDe	 ]
[InputFormat:        	org.apache.hadoop.mapred.SequenceFileInputFormat	 ]
[OutputFormat:       	org.apache.hadoop.hive.ql.io.HiveSequenceFileOutputFormat	 ]
```

Looks like the change at here somehow breaks the test (without the change, we can somehow trigger a weird code path in Hive and get at least have one column called `col`).

yhuai replied:
For now, if the schema does not have any field, can we set it to the following one?
```
[# col_name            	data_type           	comment             ]
[	 	 ]
[col                 	array<string>       	from deserializer   ]
```

So, we try to preserve the existing (and weird) behavior.

yhuai replied:
I submitted PR to try it out (https://github.com/apache/spark/pull/12363). You can find my change in the last commit (https://github.com/apache/spark/pull/12363/commits/ab70cb751cce8ca2e0757b9fa523534207864328).

andrewor14 replied:
why are there all these undocumented implicit behaviors in Hive :(

gatorsmile replied:
Now, after a few bad experiences, I think we need to add all the test cases to ensure the Hive APIs work as what we expected. See another issue I hit in `alter table ... drop partition`: https://github.com/apache/spark/pull/12220#discussion-diff-59555113

yhuai replied:
I feel it is better to have more checks in SessionCatalog to make sure that a request is valid (it can help your case).

gatorsmile replied:
Yeah, agree if you are talking about my case. : )

yhuai committed:
Preserve an existing behavior.

andrewor14 commented:
OK, I cherry-picked your changes and updated the comment.

SparkQA commented:
**[Test build #55727 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/55727/consoleFull)** for PR 12271 at commit [`55957bd`](https://github.com/apache/spark/commit/55957bdcd1dea06da750c5cf7cfa587da5250574).

SparkQA commented:
**[Test build #55727 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/55727/consoleFull)** for PR 12271 at commit [`55957bd`](https://github.com/apache/spark/commit/55957bdcd1dea06da750c5cf7cfa587da5250574).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

yhuai commented:
OK. Merging to master.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/55727/
Test PASSed.
